### PR TITLE
Core/Player rewritten duel CD reset system

### PR DESCRIPTION
**Pull request by FrancescoBorzi**
This rewrites the duel CD reset system introduced in https://github.com/TrinityCore/TrinityCore/pull/15646
- Fixed issue caused by some spells such as Stealth and Prowl, which are often cast before duel starts and broke the reset
- Now all CD are saved before the duel, reset while the duel is running, and then restored when the duel is finished.
- Tested and working

Thanks @pete318 for help.

### CODE REVIEW

File: src/server/game/Spells/SpellHistory.h [hunk 1]
```diff
@@ -117,7 +117,8 @@ class SpellHistory
     void BuildCooldownPacket(WorldPacket& data, uint8 flags, uint32 spellId, uint32 cooldown) const;
 
     CooldownStorageType::size_type GetCooldownsSizeForPacket() const { return _spellCooldowns.size(); }
-    uint16 GetArenaCooldownsSize();
+    void SaveCoolDownStateBeforeDuel();
+    void RestoreCoolDownStateAfterDuel(Player* player);
```
Comment (joschiwald) [inline src/server/game/Spells/SpellHistory.h]: useless parameter

File: src/server/game/Spells/SpellHistory.cpp [hunk 2]
```diff
@@ -637,21 +637,38 @@ void SpellHistory::BuildCooldownPacket(WorldPacket& data, uint8 flags, PacketCoo
...
-    uint16 count = 0;
+    _spellCooldownsBeforeDuel = _spellCooldowns;
+}
 
-    for (auto itr = _spellCooldowns.begin(); itr != _spellCooldowns.end();)
-    {
-        SpellInfo const* spellInfo = sSpellMgr->EnsureSpellInfo(itr->first);
+void SpellHistory::RestoreCoolDownStateAfterDuel(Player* player)
+{
+    _spellCooldowns = _spellCooldownsBeforeDuel;
```
Comment (joschiwald) [inline src/server/game/Spells/SpellHistory.cpp]: what happen if you start a profession cooldown during duell?

**Comment by Pirricli** [comment]
Is the life and mana restored too?

File: src/server/game/Spells/SpellHistory.h (hunk 1 above)
Comment (ShinDarth) [inline src/server/game/Spells/SpellHistory.h]: Player* player is used inside RestoreCoolDownStateAfterDuel method

**Comment by ShinDarth** [comment]
@Pirricli no it isn't, I can consider the idea but not in this PR

File: src/server/game/Spells/SpellHistory.h (hunk 1 above)
Comment (joschiwald) [inline src/server/game/Spells/SpellHistory.h]: use owner of the spellhistory instead

File: src/server/game/Spells/SpellHistory.cpp (hunk 2 above)
Comment (ShinDarth) [inline src/server/game/Spells/SpellHistory.cpp]: just tested and the CD remains, even if you relog

**Comment by ShinDarth** [comment]
@joschiwald 

- Tested profession CD and it's not abusable (profession CD remains even if you relog, tested both beofore duel and while duelling)

-  Remove useless player parameter

**Comment by Nyeriah** [comment]
This is very unrelated but cooldown != cool down

http://www.thefreedictionary.com/cool+down
http://www.thefreedictionary.com/cooldown

Just seen you naming things "CoolDown" so was wondering

**Comment by tkrokli** [comment]
Hehe, nice to see that I'm not the only one with a penchant for correct grammar and spelling. I suspect that it is because of the very popular initialized form "CD", even though it is not a proper acronym.

**Comment by ShinDarth** [comment]
@Nyeriah  @tkrokli thanks I didn't know it XD

**Pull request merged**